spec/unit/provider/package/dpkg_spec.rb

```diff
@@ -68,105 +70,246 @@ def dpkg_query_execution_returns(output)
       expect(Puppet::Util::Execution).to receive(:execute).with(query_args, execute_options).and_return(Puppet::Util::Execution::ProcessOutput.new(output, 0))
     end
 
+    def dpkg_query_execution_with_multiple_args_returns(output, *args)
+      args.each do |arg|
+        allow(Puppet::Util::Execution).to receive(:execute).with(arg, execute_options).and_return(Puppet::Util::Execution::ProcessOutput.new(output, 0))
+      end
+    end
+
     before do
       allow(Puppet::Util).to receive(:which).with('/usr/bin/dpkg-query').and_return(dpkgquery_path)
     end
 
     it "considers the package purged if dpkg-query fails" do
+      allow(resource).to receive(:allow_virtual?).and_return(false)
       allow(Puppet::Util::Execution).to receive(:execute).with(query_args, execute_options).and_raise(Puppet::ExecutionFailure.new("eh"))
 
       expect(provider.query[:ensure]).to eq(:purged)
     end
 
-    it "returns a hash of the found package status for an installed package" do
-      dpkg_query_execution_returns(bash_installed_output)
-
-      expect(provider.query).to eq({:ensure => "4.2-5ubuntu3", :error => "ok", :desired => "install", :name => "bash", :status => "installed", :provider => :dpkg})
-    end
-
-    it "considers the package absent if the dpkg-query result cannot be interpreted" do
-      dpkg_query_execution_returns('some-bad-data')
-
-      expect(provider.query[:ensure]).to eq(:absent)
-    end
-
-    it "fails if an error is discovered" do
-      dpkg_query_execution_returns(bash_installed_output.gsub("ok","error"))
+    context "allow_virtual true" do
+      before do
+        allow(resource).to receive(:allow_virtual?).and_return(true)
+      end
 
-      expect { provider.query }.to raise_error(Puppet::Error)
+      context "virtual_packages" do
+        let(:query_output) { 'install ok installed python 2.7.13 [python-ctypes, python-email, python-importlib, python-profiler, python-wsgiref, python-gold]' }
+        let(:virtual_packages_query_args) do
+          result = args_with_provides.dup
+          result.push(resource_name)
+        end
+
+        it "considers the package purged if dpkg-query fails" do
+          allow(Puppet::Util::Execution).to receive(:execute).with(args_with_provides, execute_options).and_raise(Puppet::ExecutionFailure.new("eh"))
+          expect(provider.query[:ensure]).to eq(:purged)
+        end
+
+        it "returns a hash of the found package status for an installed package" do
+          dpkg_query_execution_with_multiple_args_returns(query_output, args_with_provides,virtual_packages_query_args)
+          dpkg_query_execution_with_multiple_args_returns(dpkg_query_result, args, query_args)
+          expect(provider.query).to eq({:ensure => "2.7.13", :error => "ok", :desired => "install", :name => "python", :status => "installed", :provider => :dpkg})
```

Comment:
since puppet 6.0 and up dropped ruby 1.9 support, you can drop the parentheses and write it as:

```ruby
eq(ensure: "2.7.13", error: "ok", etc)
```

Which I think is more readable.